lib/runtime/src/ops/reduce.cc

```diff
@@ -45,391 +48,182 @@ ReduceParams Reduce::get_params() const {
   return params;
 }
 
-Tensor FFModel::reduce_sum(OperatorType op,
-                           const Tensor input,
-                           std::vector<int> const &_axes,
-                           bool keepdims,
-                           char const *name) {
-  Layer *rd = new Layer(this,
-                        op,
-                        DT_FLOAT,
-                        name,
-                        1 /*input*/,
-                        0 /*weights*/,
-                        1 /*outputs*/,
-                        input);
-  // Use Legion indexing to store axes
-  std::vector<int> axes;
-  for (size_t i = 0; i < _axes.size(); i++) {
-    axes.push_back(input->num_dims - 1 - _axes[i]);
-  }
-  int dims[MAX_TENSOR_DIM];
-  int numdim = input->num_dims;
-  if (keepdims) {
-    for (int i = 0; i < input->num_dims; i++) {
-      dims[i] = input->dims[i];
-    }
-    for (size_t i = 0; i < axes.size(); i++) {
-      dims[axes[i]] = 1;
-    }
-  } else {
-    numdim = 0;
-    for (int i = 0; i < input->num_dims; i++) {
-      bool reduced = false;
-      for (size_t j = 0; j < axes.size(); j++) {
-        if (axes[j] == i) {
-          reduced = true;
-        }
-      }
-      if (!reduced) {
-        dims[numdim++] = input->dims[i];
-      }
-    }
-    assert(numdim + axes.size() == input->num_dims);
-  }
-  rd->outputs[0] = create_tensor_legion_ordering(
-      numdim, dims, input->data_type, rd, 0, true /*create_grad*/);
-  rd->add_int_vector_property("legion_axes", axes);
-  rd->add_int_property("keepdims", keepdims);
-  layers.push_back(rd);
-  return rd->outputs[0];
+enum Slots {
+  INPUT,
+  OUTPUT,
+  ATTRS,
+  PROFILING,
+  REDUCE,
+  PER_DEVICE_STATE,
+  HANDLE
+};
+
+OpTaskInvocation init(TransposeAttrs const &attrs) {
+  OpTaskBinding binding;
+
+  binding.bind_arg(HANDLE, ff_handle());
+  binding.bind_arg(ATTRS, attrs);
+
+  binding.bind(INPUT, input_parallel_tensor_shape(0));
+  binding.bind(OUTPUT, output_parallel_tensor_shape(0));
+
+  return {REDUCE_INIT_TASK_ID, binding};
 }
 
-Tensor FFModel::reduce_sum(const Tensor input,
-                           std::vector<int> const &_axes,
-                           bool keepdims,
-                           char const *name) {
-  return this->reduce(OP_REDUCE_SUM, input, _axes, keepdims, name);
+static DeviceSpecific<ReducePerDeviceState>
+    init_task_impl(TaskArgumentAccessor const &acc) {
+  PerDeviceFFHandle handle = acc.get_argument<PerDeviceFFHandle>(HANDLE);
+  auto attrs = acc.get_argument<ReduceAttrs>(ATTRS);
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+
+  OperatorType = attrs.op_type;
+  // Note: How to set the reduction size?
+  size_t reduction_size ;
```

Comment:
I think `reduction_size = input.shape.get_volume() / output.shape.get_volume();`

Reply:
@lockshaw  what do you think about this?

Reply:
bump @lockshaw